go-kit/metrics/provider/otel/options.go

```diff
@@ -24,140 +23,138 @@ var (
-	}
-	return WithAttributes(attrs...)
+// Deprecated: WithExponentialHistograms is deprecated use WithExponentialAggregation instead.
+func WithExponentialHistograms() Option {
+	return WithExponentialAggregation()
+}
+
+// Deprecated: WithExplicitHistograms is deprecated use WithExplicitAggregation instead.
+func WithExplicitHistograms() Option {
+	return WithExplicitAggregation()
 }
 
-// WithServiceInstanceIDAttribute adds the "service.instance.id" attribute.
-func WithServiceInstanceIDAttribute(serviceInstanceID string) Option {
-	attrs := []attribute.KeyValue{
-		attribute.String(serviceInstanceIDKey, serviceInstanceID),
+func WithAggregationSelector(selector metric.AggregationSelector) Option {
+	return func(c *config) error {
+		c.aggregationSelector = selector
+
+		return nil
 	}
-	return WithAttributes(attrs...)
 }
 
-// WithDefaultEndpointExporter initializes the Provider with an exporter using a default endpoint.
-func WithDefaultEndpointExporter() Option {
-	return WithEndpointExporter(DefaultAgentEndpoint)
+func WithHTTPExporter(options ...otlpmetrichttp.Option) Option {
+	return WithHTTPEndpointExporter(DefaultAgentEndpoint, options...)
 }
 
-// WithEndpointExporter initializes the Provider with a default exporter.
-func WithEndpointExporter(endpoint string) Option {
-	return func(p *Provider) error {
-		if endpoint == "" {
-			return ErrEndpointNil
+func WithHTTPEndpointExporter(endpoint string, options ...otlpmetrichttp.Option) Option {
+	return WithExporterFunc(func(cfg *config) (metric.Exporter, error) {
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		defaults := []otlpmetrichttp.Option{
+			otlpmetrichttp.WithEndpoint(u.Host),
+			otlpmetrichttp.WithAggregationSelector(cfg.aggregationSelector),
+			otlpmetrichttp.WithTemporalitySelector(cfg.temporalitySelector),
 		}
-		p.exporter = defaultExporter(endpoint)
-		return nil
-	}
+
+		if u.Scheme == "https" {
+			defaults = append(defaults, otlpmetrichttp.WithTLSClientConfig(tlsconfig.New()))
+		} else {
+			defaults = append(defaults, otlpmetrichttp.WithInsecure())
+		}
+
+		if u.User.String() != "" {
+			authHeader := make(map[string]string)
+			authHeader["Authorization"] = "Basic" + base64.StdEncoding.EncodeToString([]byte(u.User.String()))
+
+			defaults = append(defaults, otlpmetrichttp.WithHeaders(authHeader))
+		}
+
+		// finally append any passed in options
+		options = append(defaults, options...)
+
+		return otlpmetrichttp.New(cfg.ctx, options...)
+	})
 }
 
-// WithExporter initializes the Provider with an exporter.
-func WithExporter(exp exporter) Option {
-	return func(p *Provider) error {
-		if exp == nil {
-			return ErrExporterNil
+func WithGRPCExporter(endpoint string, options ...otlpmetricgrpc.Option) Option {
+	return WithExporterFunc(func(cfg *config) (metric.Exporter, error) {
+		defaults := []otlpmetricgrpc.Option{
+			otlpmetricgrpc.WithEndpoint(endpoint),
+			otlpmetricgrpc.WithInsecure(),
```

Comment:
Does it have to be insecure?

Reply:
This was just copied over from the existing code. https://github.com/heroku/x/pull/201/files/c78ddf68dfd450854cc5ded6c5767429328219eb#diff-b752681e1bce242e23877d7365841220f2745ff1a7bcc49bb4457d0c34faac5aL156-L160

However, I can make it behave the same as the HTTP exporter.

Reply:
Maybe a good opportunity that could avoid surprises in the future.

Reply:
I updated the code to match the HTTP exporter. Now `http://` vs `https://` will trigger Insecure vs TLS.